server-cloud/server/auth.go

```diff
@@ -0,0 +1,199 @@
+
+	if err := sess.Save(c.Request(), c.Response()); err != nil {
+		return c.String(http.StatusInternalServerError, err.Error())
+	}
+
+	// Redirect to logged in page.
+	c.Redirect(http.StatusTemporaryRedirect, "/user")
+	return err
+}
+
+func (s *Server) logout(c echo.Context) error {
+	logoutUrl, err := url.Parse("https://" + s.conf.Auth0Domain + "/v2/logout")
+	if err != nil {
+		return c.String(http.StatusInternalServerError, err.Error())
+	}
+
+	scheme := "http"
+	if c.Request().TLS != nil {
+		scheme = "https"
+	}
+
+	fmt.Println(c.Request().Host)
+	returnTo, err := url.Parse(scheme + "://" + c.Request().Host)
+	if err != nil {
+		return c.String(http.StatusInternalServerError, err.Error())
+
+	}
+
+	parameters := url.Values{}
+	parameters.Add("returnTo", returnTo.String())
+	parameters.Add("client_id", s.conf.Auth0ClientID)
+	logoutUrl.RawQuery = parameters.Encode()
+
+	return c.Redirect(http.StatusTemporaryRedirect, logoutUrl.String())
+}
+
+func (s *Server) user(c echo.Context) error {
+	sess, err := session.Get("secret", c)
+	if err != nil {
+		fmt.Print(err)
+	}
+
+	profile := sess.Values["profile"]
+	// return c.HTML(http.StatusOK, "profile")
+	return c.Render(http.StatusOK, "user.html", profile)
+}
+
+func (s *Server) home(c echo.Context) error {
+	return c.Render(http.StatusOK, "home.html", nil)
+}
+
+func getAuth(c context.Context, conf Config) (*Authenticator, error) {
+	provider, err := oidc.NewProvider(
+		c,
+		"https://"+conf.Auth0Domain+"/",
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	config := oauth2.Config{
+		ClientID:     conf.Auth0ClientID,
+		ClientSecret: conf.Auth0ClientSecret,
+		RedirectURL:  conf.Auth0CallbackURL,
+		Endpoint:     provider.Endpoint(),
+		Scopes:       []string{oidc.ScopeOpenID, "profile"},
+	}
+
+	return &Authenticator{
+		Provider: provider,
+		Config:   config,
+	}, nil
+}
+
+func generateRandomState() (string, error) {
+	b := make([]byte, 32)
+	_, err := rand.Read(b)
+	if err != nil {
+		return "", err
+	}
+
+	state := base64.StdEncoding.EncodeToString(b)
+
+	return state, nil
+}
+
+// VerifyIDToken verifies that an *oauth2.Token is a valid *oidc.IDToken.
+func (a *Authenticator) VerifyIDToken(ctx context.Context, token *oauth2.Token) (*oidc.IDToken, error) {
+	rawIDToken, ok := token.Extra("id_token").(string)
+	if !ok {
+		return nil, errors.New("no id_token field in oauth2 token")
+	}
+
+	oidcConfig := &oidc.Config{
+		ClientID: a.ClientID,
+	}
+
+	return a.Verifier(oidcConfig).Verify(ctx, rawIDToken)
+}
+
+// IsAuthenticated is a middleware that checks if
+// the user has already been authenticated previously.
+func IsAuthenticated(next echo.HandlerFunc) echo.HandlerFunc {
+	return func(c echo.Context) error {
+		sess, err := session.Get("secret", c)
+		if err != nil {
+			fmt.Print(err)
+		}
+
+		// fmt.Println("ppp inside IsAuth", sess)
+		if sess.Values["profile"] == nil {
+			fmt.Println(" Nill profile")
+			c.Redirect(http.StatusSeeOther, "/")
+		}
```

Comment:
you mean this redirecting to / or home will be handled by frontend?, What i am doing here is once user logged out, and try to access /user again it will be automatically redirected to the /home or login page